<!DOCTYPE html>
<html>
<head>
    <title>C++/Qt</title>
    <meta charset="utf-8">
    <link rel="stylesheet" type="text/css" href="style.css">
    <link rel="stylesheet" href="fonts/font.css">
    <link href="images/Qt2.png" rel="shortcut icon" type="image/x-icon">
</head>
<body>
 <div id="site">
    <div id="header">
       <div class="logo"><center><a href="C++.html"  name="C++" title="C++"><img src="images/logo_c.svg" class="logoimg"></a></center></div>
        <div class="logo2"><center><a href="Qt.html"  name="Qt" title="Qt"><img src="images/logo_qt.png" class="logoimg2"></a></center></div>
        <div class="title"><center><a href="index.html"  name="main" title="Головна"><img src="images/title.png" class="titleimg"></a></center></div>
        
    </div>
    <div id="content">
        <div id="menu3">
            <ul>
                <li><a href="type.html">Типи даних</a>
                </li>
                <li><a href="operators.html">Оператори й операції</a>
                </li>
                <li><a href="managing_structures.html">Керуючі структури</a>
                </li>
                <li><a href="cycle.html">Цикли</a>
                </li>
                <li><a href="functions.html">Функції</a>
                </li>
                <li><a href="standart_functions.html">Стандартні Функції</a>
                </li>
                <li><a href="array.html">Масиви []</a>
                </li>
                <li><a href="ptr.html">Вказівники *, Силки&</a>
                </li>
                <li><a href="string.html">Робота зі стрічками</a>
                </li>
                <li><a href="struct.html">Структури</a>
                </li>
                <li><a href="dynamic_struct.html">Динамічні Структури даних</a>
                </li>
                <li><a href="fstream.html">Робота з файлами(Fstream)</a>
                </li>
                <li><a href="template.html">Шаблони функцій</a>
                </li>
                <li><a href="clases.html">ООП Класи</a>
                </li>
                <li><a href="constructors.html">Конструктор і деструктор</a>
                </li>
                <li><a href="peregruzka.html">Перегрузка операцій</a>
                </li>
                <li><a href="nasliduvanya.html">Наслідування</a>
                </li>
                <li><a href="virtual.html">Віртуальні функції</a>
                </li>
                <li><a href="exception.html">Винятки</a>
                </li>
                <li><a href="stl.html">STL</a>
                </li>
            </ul>
       </div>
        <div id="menu4">
            <ul>
                <li><a href="qt_creator.html">Qt Creator</a>
                </li>
                <li><a href="Widgets.html">Віджети та компонування</a>
                </li>
                <li><a href="comp.html">Класи компонування</a>
                </li>
                <li><a href="Modul.html">Модулі Qt</a>
                </li>
                <li><a href="qt_designer.html">Знайомство з Qt Designer</a>
                </li>
                <li><a href="signals.html">Сигнали и слоти</a>
                </li>
                <li><a href="mainwindow.html">Класи главного вікна</a>
                </li>
                </li>
                <li><a href="dialog.html">Діалогові вікна</a>
                </li>
                <li><a href="Modal.html">Модальні немодальні вікна</a>
                </li>
                <li><a href="Menu.html">Створення меню в Qt</a>
                </li>
                <li><a href="qfile.html">Робота з файлами</a>
                </li>
                <li><a href="qdata.html">Клас дати QDate</a>
                </li>
                <li><a href="qtime.html">Клас часу QTime</a>
                </li>

                <li><a href="conteiner.html">Контейнери в Qt</a>
                </li>
                <li><a href="iterator.html">Ітератори</a>
                </li>
                <li><a href="foreach.html">Конструкція foreach</a>
                </li>
                <li><a href="Webkit.html">QtWebKit Guide</a>
                </li>
                <li><a href="sql.html">Модуль QtSql</a>
                </li>
                <li><a href="paint.html">Система малювання</a>
                </li>
                <li><a href="test.html">Тестування в Qt</a>
                </li>
            </ul>
       </div> 
        <div id="wrapper">
             <center><h1>exception в С++</h1></center>
        <br>
        <p>Всім нам інтуїтивно зрозуміло що таке виключення. Це якась дія, яка набирає чинності, при настанні нестандартної ситуації. Говорячи про про винятки в С ++, розглянемо одну з таких нестандартних ситуацій - поділ на 0. Уявіть, що ми працюємо в програмі, де вручну вносимо багато числових даних. Далі, в результаті якихось розрахунків програми, значення однієї із змінних стане рівним 0. А в наступному рядку коду це значення виступає дільником. Звичайно, ця дія призведе до неминучого закриття програми і виходу з неї. І ви можете собі уявити свій стан, коли півгодини вводите дані, працюючи в цій програмі, і на якомусь етапі вона просто перестає працювати і закривається, а всі дані пропадають.</p>
        <p>У цій статті буде наведено дуже простий і короткий приклад, який допоможе розібратися з винятками в С ++. У ньому користувачеві буде запропоновано ввести два числа і програма проведе прості розрахунки з цими числами. Введення і розрахунки будуть проводитися певна кількість раз - за допомогою циклу while.</p>
        <p>Спочатку наберіть вихідний код нижче і скомпілюйте. У змінну num2 внесіть число 0, щоб у вас було розуміння тієї проблеми, що якщо в програмі зустрічається поділ на 0, вона різко припиняє роботу, хочете ви цього чи ні.</p>
        <p><ex class="ex">&nbsp;&nbsp; 1. #include &lt;iostream>; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 2. using namespace std; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 3. int main() &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 4. { &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 5. setlocale(LC_ALL, "ukr"); &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 6. int num1; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 7. int num2; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 8. int var = 2; // керуюча змінна для while &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 9. while(var--) // поки var - істина (не дорівнює 0) &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 10. { &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 11. cout &lt;&lt; "Введіть значення num1: "; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 12. cin >> num1; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 13. cout &lt;&lt; "Введіть значення num2: "; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 14. cin >> num2; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 15. cout &lt;&lt; "num1 + num2 = " &lt;&lt; num1 + num2 &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 16. cout &lt;&lt; "num1 / num2 = " &lt;&lt; num1 / num2 &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 17. cout &lt;&lt; "num1 - num2 = " &lt;&lt; num1 - num2 &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 18. cout &lt;&lt; "=================================" &lt;&lt; endl &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 19. } &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 20. cout &lt;&lt; "Программа завершила работу!" &lt;&lt; endl &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 21. return 0; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 22. } &nbsp;&nbsp;</ex></p>
        <p>У програмі цикл повинен відпрацювати два рази. У рядку 9 ми передаємо циклу значення змінної var і в цьому ж рядку зменшуємо її на одиницю, використовуючи операцію декремента (var--). Але, якщо в змінну num2 буде введено значення 0, ми побачимо тільки результат складання a + b і програма завершить роботу не залежно від того, є код нижче чи ні, відпрацював цикл чи ні.</p>
        <p>Використовуючи виключення, ми зможемо уникнути таких проблем. Щоб «прикрутити» виняток до цього прикладу, треба познайомитися з наступними командами С ++: throw (в перекладі - обробити, запустити), try (спроба), catch (зловити, ловити) .В вихідному коді нижче, виняток спрацює так: програма отримує конкретне вказівку від програміста - якщо значення певної змінної в певній ділянці коду (в try-блоці) дорівнюватиме 0, то в цьому випадку нехай генерується виключення throw. Це виняток автоматично передасться catch-блоку у вигляді параметра і виконається код цього блоку.</p>
        <p>Відразу розглянемо приклад. А все докладні пояснення по вихідного коду і теорія будуть розташовуватися під ним:</p>
        <p><ex class="ex">&nbsp;&nbsp; 1. #include &lt;iostream>; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 2. using namespace std; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 3. int main() &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 4. { &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 5. setlocale(LC_ALL, "ukr"); &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 6. int num1; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 7. int num2; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 8. int var = 2; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 9. while(var--) &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 10. { &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 11. cout &lt;&lt; "Введіть значення num1: "; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 12. cin >> num1; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 13. cout &lt;&lt; "Введіть значення num2: "; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 14. cin >> num2; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 15. cout &lt;&lt; "num1 + num2 = " &lt;&lt; num1 + num2 &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 16. cout &lt;&lt; "num1 / num2 = "; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 17. try // код, який може привести до помилки, розташовується тут &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 18. { &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 19. if (num2 == 0) &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 20. { &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 21. throw 123; //генерувати ціле число 123 &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 22. } &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 23. cout &lt;&lt; num1 / num2 &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 24. } &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 25. catch(int i)//сюда передається число 123 &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 26. { &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 27. cout &lt;&lt; "Помилка №" &lt;&lt; i &lt;&lt; " - на 0 делить нельзя!!!!" &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 28. } &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 29. cout &lt;&lt; "num1 - num2 = " &lt;&lt; num1 - num2 &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 30. cout &lt;&lt; "=================================" &lt;&lt; endl &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 31. } &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 32. cout &lt;&lt; "Программа завершила работу!" &lt;&lt; endl &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 33. return 0; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 34. } &nbsp;&nbsp;</ex></p>
        <p>У рядках 17 - 24 визначено try-блок. У ньому розташовується код, який потенційно може викликати помилку в роботі програми, а саме помилку в разі поділу на 0. Задаємо умова if - якщо num2 дорівнює 0, то генерувати ціле число 123, наприклад. В цьому випадку try-блок відразу припиняє виконання подальших команд, а число 123 «падає» в catch. У нашому прикладі він виводить повідомлення про помилку. При цьому програма продовжує працювати і виконувати команди, розміщені нижче. Якщо ж число num2 не дорівнюватиме нулю, то в try-блоці виконається команда cout << num1 / num2 << endl; , А catch не спрацює.</p>
        <p>Щоб краще зрозуміти, як саме значення, яке генерує throw передається в catch, спробуйте замінити ділянку коду рядків 22 - 33 на наступний код:</p>
        <p><ex class="ex">&nbsp;&nbsp; 1. try &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 2. { &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 3. if (num2 == 0) &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 4. {&nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 5. throw "Помилка - на 0 Ділити не можна!!!!"; // генерувати символьний рядок &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 6. }&nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 7. cout &lt;&lt; num1 / num2 &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 8. }&nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 9. catch(char *str) //Сюда передається стрічка&nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 10. { &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 11. cout &lt;&lt; str &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 12. } &nbsp;&nbsp;</ex></p>
        <p>Тут при спробі поділу на 0, throw генерується не число, а символьний рядок. І блоку catch в дужках ми показуємо, що він буде приймати покажчик на рядок і нижче - виводити цей рядок на екран.</p>
        <p>Виняток так само може генеруватися в певній програмістом функції. Наприклад у нас визначена функція розподілу двох цілих чисел і в ній ми відразу перевіримо розподіл на 0.</p>
        <p><ex class="ex">&nbsp;&nbsp; 1. int division(int n1, int n2) &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 2. { &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 3. if (n2 == 0) &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 4. { &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 5. throw 99; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 6. } &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 7. return n1 / n2; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 8. } &nbsp;&nbsp;</ex></p>
        <p>А викликати цю функцію, звичайно треба в try-блоці.</p>
        <p><ex class="ex">&nbsp;&nbsp; 1. try &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 2. { &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 3. cout &lt;&lt; division(5, 0); &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 4. cout &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 5. } &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 6. catch(int i) &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 7. { &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 8.  cout &lt;&lt; "Помилка №" &lt;&lt; i &lt;&lt; " Ділення на ноль." &lt;&lt; endl; &nbsp;&nbsp;</ex></p>
        <p><ex class="ex">&nbsp;&nbsp; 9. } &nbsp;&nbsp;</ex></p> 
        <p>Ділення на нуль не станеться й catch зловить число 99.</p>
        <p>Що важливо запам'ятати про винятки:</p>
        <p>1. try-блок - так званий блок повторних спроб. У ньому треба розташовувати код, який може привести до помилки і аварійного закриття програми;</p>
        <p>2. throw генерує виняток. Те що зупинить роботу try-блоку і призведе до виконання коду catch-блоку. Тип винятку повинен відповідати, типу прийнятого аргументу catch-блоку;</p>
        <p>3. catch-блок - вловлює блок, зловить то, що визначив throw і виконає свій код. Цей блок повинен розташовуватися безпосередньо під try-блоком. Ніякої код не повинен їх розділяти.</p>
        <p>4. якщо в try-блоці винятку не генерувалося, catch-блок не спрацює. Програма його обійде.</p>
        </div>
    </div>
    <div id="footer">
       <div id="contacts"><center><h2>Контакти<hr></h2>
           <a href="http://www.facebook.com/zubykandriy" target="blank"><img src="images/Facebooklogo2.png" class="logo_of_contacts" alt="Facebook" title="Facebook"></a><br>  
           <a href="https://vk.com/zubyk.andriy" target="blank"><img src="images/vk2.png" class="logo_of_contacts" alt="Vk" title="Vk"></a></center>
       </div>
       <div id="other"><center><h2>IDE<hr></h2>
       <a href="https://www.visualstudio.com/ru/downloads/?rr=https%3A%2F%2Fwww.google.com.ua%2F" target="blank"><img src="images/VS.png" class="logo_of_other" alt="Visual Studio" title="Visual Studio"></a><br>
       <a href="https://www.qt.io/download" target="blank"><img src="images/Qt2.png" class="logo_of_other" alt="Qt Creator" title="Qt Creator"></a></center>
       </div>
       <div id="signature"><center><p>Зубик Андрій &copy; 2018<br><span class="hp"><img src="images/logo_c.svg"> Довідник С++/Qt <img src="images/logo_qt2.png"></span></p></center></div>
    </div>
 </div>
</body>
</html>